jinja2: improve docs for globals, filters and tests (#958)
cylc/cylc-doc: src/user-guide/writing-workflows/jinja2.rst

      {{ CYLC_WORKFLOW_RUN_DIR | default("not-defined") }}

.. _Jinja2Filters:

Jinja2 Filters, Tests and Globals
---------------------------------

.. _Jinja2 Built-in Globals: https://jinja.palletsprojects.com/en/stable/templates/#list-of-global-functions
.. _Jinja2 Built-in Filters: https://jinja.palletsprojects.com/en/stable/templates/#list-of-builtin-filters
.. _Jinja2 Built-in Tests: https://jinja.palletsprojects.com/en/stable/templates/#builtin-tests

Jinja2 provides "globals", "filters" and "tests" which can be helpful in
workflow writing.

Globals

>    Regular Python functions.

Reviewer: ```suggestion
   Regular Python variables and functions.
```

Author: Do we want to advertise this?

Reviewer: I'm not exactly sure what the use cases are, but can imagine something similar to our `environ` variable that we provide as a global?

Author: I can't think of anything good at present.

A worry is that folk will attempt to use this for data which is a dubious pattern with caveats.

Reviewer: Maybe warn against such a pattern in that case? As is, potentially confusing referring to them exclusively as functions when it goes on to list `environ` as a built-in global.

   :Cylc builtins: :ref:`user-guide.jinja2.cylc-builtin-globals`
   :Custom directory: :ref:`Jinja2Globals <user-guide.jinja2.custom-extensions>`
Filters
   Special functions which "chain" using the pipe character (``|``).

   :Jinja2 builtins: `Jinja2 Built-in Filters`_
   :Cylc builtins: :ref:`user-guide.jinja2.cylc-builtin-filters`
   :Custom directory: :ref:`Jinja2Filters <user-guide.jinja2.custom-extensions>`
Tests
   Special functions which work with the ``is`` operator.

   :Jinja2 builtins: `Jinja2 Built-in Tests`_
   :Custom directory: :ref:`Jinja2Tests <user-guide.jinja2.custom-extensions>`

For example, this :cylc:conf:`flow.cylc` file uses the
:py:func:`pad <cylc.flow.jinja.filters.pad.pad>` filter to help write out
task definitions:

.. code-block:: cylc

   [runtime]
   {% for x in range(3) %}
       [[task_{{ x | pad(3) }}]]
           script = sleep {{ x }}
   {% endfor %}

The Jinja2 would be expanded like so:

.. code-block:: cylc

   [runtime]
       [[x_001]]
           script = sleep 1
       [[x_002]]
           script = sleep 2
       [[x_003]]
           script = sleep 3

In addition to the built-in's that Jinja2 and Cylc provide, we can also define

> our own custom filters (see :ref:`user-guide.jinja2.custom-extensions`).

Reviewer: ```suggestion
In addition to the built-ins that Jinja2 and Cylc provide, you can also define
your own custom filters (see :ref:`user-guide.jinja2.custom-extensions`).
```

.. _user-guide.jinja2.cylc-builtin-globals:

Cylc Built-in Globals
^^^^^^^^^^^^^^^^^^^^^

.. list-table::

   * - :py:data:`environ`
     - Access environment variables.
   * - :py:func:`raise`
     - Raise an error.
   * - :py:func:`assert`
     - Raise an error if a condition is not met.

.. _jinja2-environ:

.. py:data:: environ

   Provides access to environment variables.

   Note, these are the "parse-time" environment variables - i.e, the environment

>    that is set when the workflow's :cylc:conf`flow.cylc` file is processed.

Reviewer: ```suggestion
   that is set when the workflow's :cylc:conf:`flow.cylc` file is processed.
```

   This happens when a workflow is validated or started, not when jobs are
   submitted. Jinja2 does not have access to dynamic environment variables
   available to jobs.

   .. describe:: Jinja2 Examples:

      .. code-block:: cylc

         [runtime]
             [[root]]
                 [[[environment]]]
                     HOME_DIR_ON_WORKFLOW_HOST = {{environ['HOME']}}

.. _jinja2-raise:

.. py:function:: raise(error_message)

   The ``raise`` function will result in an error containing the provided text.

   Calling this will cause ``cylc validate`` to fail with the provided error
   message and will prevent the workflow from being started. It's useful for
   validating input template variables.

   .. describe:: Jinja2 Examples:

      .. code-block:: cylc

>          {% if not VARIABLE is defined %}

Reviewer: ```suggestion
         {% if VARIABLE is not defined %}
```

             {{ raise('VARIABLE must be defined for this workflow.') }}
         {% endif %}

.. _jinja2-assert:

> .. py:function:: assert(codition, error_message)

Reviewer: ```suggestion
.. py:function:: assert(condition, error_message)
```

.. autofunction:: cylc.flow.jinja.filters.duration_as.duration_as

.. _CustomJinja2Filters:
.. _user-guide.jinja2.custom-extensions:

Custom Jinja2 Extensions
^^^^^^^^^^^^^^^^^^^^^^^^

> Custom Jinja2 globals, filters and tests can be defined within workflows.

Reviewer: ```suggestion
You can define your own custom Jinja2 globals, filters and tests within workflows.
```

> These extensions are Python modules containing a function with the same name

Reviewer: ```suggestion
These extensions must be Python modules containing a function with the same name
```

as the module (e.g, a module called ``foo.py`` should contain a function called
``foo``).

Jinja2 globals, go in the workflow :term:`source directory` in a subdirectory

> called ``Jinja2Globals``, filters into ``Jinja2Filters`` and tests into

Reviewer: ```suggestion
Jinja2 globals go in the workflow :term:`source directory` in a subdirectory
called ``Jinja2Globals``, filters in ``Jinja2Filters`` and tests in
```

``Jinja2Tests``.

This example defines one of each and demonstrates how to use them:

.. code-block:: cylc
   :caption: flow.cylc

   #!Jinja2

>    # "globals" are regular Python functions

Reviewer: ```suggestion
   # "globals" are regular Python functions or variables
```
